typing/typecore.ml

```diff
@@ -4144,8 +4144,9 @@ and type_argument ?explanation ?recarg env sarg ty_expected' ty_expected =
     | Pexp_ifthenelse (_, e1, Some e2) -> is_inferred e1 && is_inferred e2
     | _ -> false
   in
-  match expand_head env ty_expected' with
-    {desc = Tarrow(Nolabel,ty_arg,ty_res,_); level = lv}
+  let lv = (repr ty_expected').level in
+  match expand_head env (correct_levels ty_expected') with
```

Comment:
The appearance of correct_levels makes me nervous. I tend to think that we should never really need to use it -- the levels should already be correct. Could you add a comment explaining why it is needed here?

Reply:
The other name of `correct_levels` is `duplicate_type`.
Here this allows to check the expansion of `ty_expected'` without changing the scopes in the original type.
Otherwise we get spurious ambivalence, even though the type doesn't have an interesting shape. We then expand the real type if we know that it is going to be useful.

Note that all this means that there is a high potential that this PR as a whole introduces spurious ambivalence in other similar places. I should probably find a better way to do it, and review all uses of `expand_head`.